src/fcntl.rs

```diff
@@ -596,39 +596,124 @@ pub fn fcntl(fd: RawFd, arg: FcntlArg) -> Result<c_int> {
     Errno::result(res)
 }
 
-// TODO: convert to libc_enum
 #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
 #[non_exhaustive]
 pub enum FlockArg {
     LockShared,
     LockExclusive,
-    Unlock,
     LockSharedNonblock,
     LockExclusiveNonblock,
-    UnlockNonblock,
 }
 
+/// Represents valid types for flock.
+///
+/// # Safety
+/// Types implementing this must be `!Clone`.
 #[cfg(not(any(target_os = "redox", target_os = "solaris")))]
-pub fn flock(fd: RawFd, arg: FlockArg) -> Result<()> {
-    use self::FlockArg::*;
+pub unsafe trait Flockable: AsRawFd {}
 
-    let res = unsafe {
-        match arg {
-            LockShared => libc::flock(fd, libc::LOCK_SH),
-            LockExclusive => libc::flock(fd, libc::LOCK_EX),
-            Unlock => libc::flock(fd, libc::LOCK_UN),
-            LockSharedNonblock => {
-                libc::flock(fd, libc::LOCK_SH | libc::LOCK_NB)
-            }
-            LockExclusiveNonblock => {
-                libc::flock(fd, libc::LOCK_EX | libc::LOCK_NB)
-            }
-            UnlockNonblock => libc::flock(fd, libc::LOCK_UN | libc::LOCK_NB),
+/// Represents an owned flock, which unlocks on drop.
+///
+/// See flock(2) for details on locking semantics.
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+#[derive(Debug)]
+pub struct Flock<T: Flockable>(Option<T>);
```

Comment:
Why Option?  When will it ever be None?

Reply:
None is used here so that we can .take() the inner type out of the lock when calling Flock.unlock()

Reply:
It will only ever be none while it is in the process of dropping, and only when after a successful `unlock()` call. It prevents a double-free situation; see above conversations for more details.

Reply:
I don't like Option because it adds extra runtime code.  You can do it without Option like this:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5713996199c6d48d422690bb50fda583

Reply:
Nice! `std::mem::forget()` is the way to go!